Source/WebCore/accessibility/AccessibilityObject.cpp

```diff
@@ -1768,26 +1768,28 @@ String AccessibilityObject::stringForVisiblePositionRange(const VisiblePositionR
     return builder.toString();
 }
 
-VisiblePosition AccessibilityObject::nextLineEndPosition(const VisiblePosition& visiblePos) const
+VisiblePosition AccessibilityObject::nextLineEndPosition(const VisiblePosition& startPosition) const
 {
-    if (visiblePos.isNull())
-        return VisiblePosition();
-
-    // to make sure we move off of a line end
-    VisiblePosition nextVisiblePos = visiblePos.next();
-    if (nextVisiblePos.isNull())
-        return VisiblePosition();
+    if (startPosition.isNull())
+        return { };
 
-    VisiblePosition endPosition = endOfLine(nextVisiblePos);
+    // Move to the next position to ensure we move off a line end.
+    auto nextPosition = startPosition.next();
+    if (nextPosition.isNull())
+        return { };
 
-    // as long as the position hasn't reached the end of the doc,  keep searching for a valid line end position
-    // There are cases like when the position is next to a floating object that'll return null for end of line. This code will avoid returning null.
-    while (endPosition.isNull() && nextVisiblePos.isNotNull()) {
-        nextVisiblePos = nextVisiblePos.next();
-        endPosition = endOfLine(nextVisiblePos);
+    auto lineEndPosition = endOfLine(nextPosition);
+    // As long as the position hasn't reached the end of the document, keep searching for a valid line
+    // end position. Skip past null positions, as there are cases like when the position is next to a
+    // floating object that'll return null for end of line. Also, in certain scenarios, like when one
+    // position is editable and the other isn't (e.g. in mixed-contenteditable-visible-character-range-hang.html),
+    // we may end up back at the same position we started at. This is never valid, so keep moving forward
+    // trying to find the next line end.
+    while ((lineEndPosition.isNull() || lineEndPosition == startPosition) && nextPosition.isNotNull()) {
+        nextPosition = nextPosition.next();
+        lineEndPosition = endOfLine(nextPosition);
     }
-
-    return endPosition;
+    return lineEndPosition;
 }
 
 std::optional<VisiblePosition> AccessibilityObject::previousLineStartPositionInternal(const VisiblePosition& visiblePosition) const
```

Comment:
Does this account for the case where you have two or more <br>s together?
If in the case of a line ending in a non-editable space followed by <br>, we return the <br> as the end of line position instead of the space, would the hang occur? I believe returning the <br> position is the right thing because what the client wants is the line range, and the end of that range should be past the last character.

Reply:
> If in the case of a line ending in a non-editable space followed by a br, we return the br as the end of line position instead of the space, would the hang occur?

It would, as the `<br>` is also wrapped by `contenteditable="false"`.

> Does this account for the case where you have two or more `br`s together?

I believe so, but I can add another testcase to be certain.

>  I believe returning the `br` position is the right thing because what the client wants is the line range, and the end of that range should be past the last character.

That makes sense. I think in general, this PR isn't changing the algorithm around what is the line end for any given position, just making sure we don't get stuck. I have tested this change on Google docs and VoiceOver output sounds good, so I think we should wait for a motivating example to investigate this bit further.